From c11a166cd4c19664355e0e3d9c04cfa7ee4aa9f4 Mon Sep 17 00:00:00 2001
From: Colin Cross <ccross@android.com>
Date: Thu, 15 Apr 2010 15:21:51 -0700
Subject: android-common: Fix slab.h includes for 2.6.34-rc4
Patch-mainline: HEAD
Git-commit: c11a166cd4c19664355e0e3d9c04cfa7ee4aa9f4

Signed-off-by: Greg Kroah-Hartman <gregkh@suse.de>

diff --git a/drivers/staging/android/binder.c b/drivers/staging/android/binder.c
index 99010d4..3aec4530 100644
--- a/drivers/staging/android/binder.c
+++ b/drivers/staging/android/binder.c
@@ -31,6 +31,7 @@
 #include <linux/sched.h>
 #include <linux/uaccess.h>
 #include <linux/vmalloc.h>
+#include <linux/slab.h>
 
 #include "binder.h"
 
diff --git a/drivers/staging/android/logger.c b/drivers/staging/android/logger.c
index 64cc2a1..dd76564 100644
--- a/drivers/staging/android/logger.c
+++ b/drivers/staging/android/logger.c
@@ -23,6 +23,7 @@
 #include <linux/miscdevice.h>
 #include <linux/uaccess.h>
 #include <linux/poll.h>
+#include <linux/slab.h>
 #include <linux/time.h>
 #include "logger.h"
 
